### PR TITLE
Use non-atomic flushing with block replay

### PR DESCRIPTION
**Pull request by furszy**
> This patch adds an extra "head blocks" to the chainstate, which gives the range of blocks for writes may be incomplete. At the start of a flush, we write this record, write the dirty dbcache entries in 16 MiB batches, and at the end we remove the heads record again. If it is present at startup it means we crashed during flush, and we rollback/roll forward blocks inside of it to get a consistent tip on disk before proceeding.

> If a flush completes succesfully, the resulting database is compatible with previous versions. If the node crashes in the middle of a flush, a version of the code with this patch is needed to recovery.

An adaptation of the following PRs with further modifications to the `feature_dbcrash.py` test to be up-to-date with upstream and solve RPC related bugs.

* #10148.
* Increase RPC wait time.
* #11831
* #11593
* #12366
* #13837
* #13894

### CODE REVIEW

**Comment by furszy** [comment]
Added two more commits solving the RPC timeout, GA should be good now. Ready for review.

**Review by random-zebra — COMMENTED** [review]
Good stuff. Code review ACK with some points.

File: src/init.cpp [hunk 1]
```diff
@@ -1612,6 +1616,13 @@ bool AppInitMain()
                     break;
                 }
 
+                if (!ReplayBlocks(chainparams, pcoinsdbview)) {
+                    strLoadError = _("Unable to replay blocks. You will need to rebuild the database using -reindex-chainstate.");
```
[the remark this answers is not in the document]
Comment (furszy) [reply src/init.cpp:1622]: k yeah, that is a good PR that we could revive soon.

File: src/chain.cpp [hunk 2]
```diff
@@ -326,8 +326,24 @@ bool CBlockIndex::RaiseValidity(enum BlockStatus nUpTo)
     return false;
 }
 
-/*
- * CBlockIndex - Legacy Zerocoin
- */
+/** Find the last common ancestor two blocks have.
+ *  Both pa and pb must be non-NULL. */
+CBlockIndex* LastCommonAncestor(CBlockIndex* pa, CBlockIndex* pb)
```
[the remark this answers is not in the document]
Comment (furszy) [reply src/chain.cpp:331]: I deliberately left it in this way, need some other changes like move `CNodeState::pindexLastCommonBlock` and  `CNodeState::pindexBestKnownBlock` to const as well.

File: src/chain.h [hunk 3]
```diff
@@ -284,6 +284,9 @@ class CBlockIndex
     const CBlockIndex* GetAncestor(int height) const;
 };
 
+/** Find the forking point between two chain tips. */
+CBlockIndex* LastCommonAncestor(CBlockIndex* pa, CBlockIndex* pb);
```
[the remark this answers is not in the document]
Comment (furszy) [reply src/chain.h:288]: same as above https://github.com/PIVX-Project/PIVX/pull/2207#discussion_r579341337 .

File: src/validation.cpp [hunk 4]
```diff
@@ -3703,9 +3703,9 @@ bool ReplayBlocks(const CChainParams& params, CCoinsView* view)
     uiInterface.ShowProgress(_("Replaying blocks..."), 0);
     LogPrintf("Replaying blocks\n");
 
-    const CBlockIndex* pindexOld = nullptr;  // Old tip during the interrupted flush.
-    const CBlockIndex* pindexNew;            // New tip during the interrupted flush.
-    const CBlockIndex* pindexFork = nullptr; // Latest block common to both the old and the new tip.
+    CBlockIndex* pindexOld = nullptr;  // Old tip during the interrupted flush.
+    CBlockIndex* pindexNew;            // New tip during the interrupted flush.
+    CBlockIndex* pindexFork = nullptr; // Latest block common to both the old and the new tip.
```
[the remark this answers is not in the document]
Comment (furszy) [reply src/validation.cpp]: Saw this comment after writing in the other two.. all good, will cherry-pick the commit 👍 .

**Comment by furszy** [comment]
Done @random-zebra, commit cherry-picked.

File: src/validation.cpp [hunk 5]
```diff
@@ -3663,6 +3670,92 @@ bool CVerifyDB::VerifyDB(CCoinsView* coinsview, int nCheckLevel, int nCheckDepth
...
+    if (mapBlockIndex.count(hashHeads[0]) == 0) {
+        return error("ReplayBlocks(): reorganization to unknown block requested");
+    }
+    pindexNew = mapBlockIndex[hashHeads[0]];
+
+    if (!hashHeads[1].IsNull()) { // The old tip is allowed to be 0, indicating it's the first flush.
+        if (mapBlockIndex.count(hashHeads[1]) == 0) {
+            return error("ReplayBlocks(): reorganization from unknown block requested");
+        }
+        pindexOld = mapBlockIndex[hashHeads[1]];
```
[the remark this answers is not in the document]
Comment (furszy) [reply src/validation.cpp]: pushed in 45aa5b7.

**Review by random-zebra — APPROVED** [review]
ACK aab15d7eb4a358a0cff2d4f8df1af13a12cc4739

**Review by Fuzzbawls — APPROVED** [review]
ACK aab15d7eb4a358a0cff2d4f8df1af13a12cc4739

**Pull request merged**